File: src/c47/solver/sumprod.c
/* This file is part of 43S.
 *
 * 43S is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 43S is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 43S.  If not, see <http://www.gnu.org/licenses/>.
 */

/********************************************//**
 * \file sumprod.c
 ***********************************************/

#include "solver/sumprod.h"

#include "c43Extensions/addons.h"
#include "constantPointers.h"
#include "defines.h"
#include "display.h"
#include "error.h"
#include "flags.h"
#include "fonts.h"
#include "integers.h"
#include "items.h"
#include "mathematics/comparisonReals.h"
#include "mathematics/integerPart.h"
#include "mathematics/iteration.h"
#include "mathematics/multiplication.h"
#include "programming/lblGtoXeq.h"
#include "programming/manage.h"
#include "programming/nextStep.h"
#include "realType.h"
#include "registers.h"
#include "registerValueConversions.h"
#include "screen.h"
#include "solver/solve.h"
#include "statusBar.h"
#include "stack.h"
#include "timer.h"
#include "c47.h"

#include <string.h>


//Complex operation:
//  The counter is always Real34.
//  The result of f(n) can be complex, in which case if CPXRES is set, operation continues in complex.
//  If not set, an error is raised.



#if !defined(TESTSUITE_BUILD)

  void showProgressReal(const real_t *a, real_t *ai, bool_t cpx) {
    real34_t a34, ai34;
    #if ENABLE_SOLVER_PROGRESS == 1
        uint8_t savedDisplayFormatDigits = displayFormatDigits;

        clearRegisterLine(REGISTER_Z, true, true);
        clearRegisterLine(REGISTER_Y, true, true);
        clearRegisterLine(REGISTER_X, true, true);

        displayFormatDigits = displayFormat == DF_ALL ? 0 : 33;
        if(!cpx) {
          realToReal34(a, &a34);
          real34ToDisplayString(&a34, amNone, tmpString, &standardFont, 9999, 34, false, true);
          showString(tmpString, &standardFont, 1, Y_POSITION_OF_REGISTER_X_LINE + 6, vmNormal, true, true);
        } else {
          realToReal34(a, &a34);
          realToReal34(ai, &ai34);
          real34ToDisplayString(&a34, amNone, tmpString, &standardFont, 9999, 34, false, true);
          showString(tmpString, &standardFont, 1, Y_POSITION_OF_REGISTER_Y_LINE + 6, vmNormal, true, true);
          uint32_t x = 0;
          if(real34CompareGreaterEqual(&ai34,const34_0)) {
            x = showString("+", &standardFont, 1, Y_POSITION_OF_REGISTER_X_LINE + 6, vmNormal, true, true);
          } else {
            x = showString(" ", &standardFont, 1, Y_POSITION_OF_REGISTER_X_LINE + 6, vmNormal, true, true);
          }
          real34ToDisplayString(&ai34, amNone, tmpString, &standardFont, 9999, 34, false, true);
          strcat(tmpString, COMPLEX_UNIT);
          showString(tmpString, &standardFont, x, Y_POSITION_OF_REGISTER_X_LINE + 6, vmNormal, true, true);
        }
        displayFormatDigits = savedDisplayFormatDigits;

      #if defined DMCP_BUILD
        lcd_refresh();
      #endif //DMCP_BUILD
    #endif // ENABLE_SOLVER_PROGRESS == 1
  }


  static void _programmableSumProd(uint16_t label, bool_t prod) {
    currentKeyCode = 255;
    int32_t       loop = 0;
    int16_t       finished = 0;
    bool_t        abort = false;
    real_t        resultX, resultXi, resultR, resultRi;
    real34_t      loopStep, loopTo, counter, compare, sign, rLoop;
    bool_t        changedOverToComplex = false;
    longInteger_t iLoop;
    fnToReal(NOPARAM);
    real34Copy(REGISTER_REAL34_DATA(REGISTER_X), &loopStep);
    fnDrop(NOPARAM);
    fnToReal(NOPARAM);
    real34Copy(REGISTER_REAL34_DATA(REGISTER_X), &loopTo);
    fnDrop(NOPARAM);
    fnToReal(NOPARAM);
    real34Copy(REGISTER_REAL34_DATA(REGISTER_X), &counter); //Loopfrom
    realCopy(prod ? const_1 : const_0, &resultR);           //Initialize real accumulator
    realCopy(const_0, &resultRi);                           //Initialize complex accumulator

    real34Subtract(&loopTo, &counter, &rLoop);              //calculate the remaining iteration counter
    if(!real34IsZero(&loopStep)) {
      real34Divide(&rLoop, &loopStep, &rLoop);
    }
    longIntegerInit(iLoop);
    convertReal34ToLongInteger(&rLoop, iLoop, DEC_ROUND_DOWN);
    loop = (int32_t)longIntegerModuloUInt(iLoop, (int32_t)(0x7FFFFFFF));
    longIntegerFree(iLoop);

    if( !real34CompareEqual(&loopTo, &counter) &&
        (  real34IsZero(&loopStep) ||
          (real34CompareGreaterThan(&loopTo, &counter) && real34CompareLessEqual(&loopStep, const34_0)) ||
          (real34CompareLessThan(&loopTo, &counter) && real34CompareGreaterEqual(&loopStep, const34_0))
        )
      ) {
      displayCalcErrorMessage(ERROR_BAD_INPUT, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        sprintf(errorMessage, "Counter will not count to destination");
        moreInfoOnError("In function _programmableiSumProd:", errorMessage, NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    }
    else {
      ++currentSolverNestingDepth;
      setSystemFlag(FLAG_SOLVING);

      while(lastErrorCode == ERROR_NONE) {
        hourGlassIconEnabled = true;
        showHideHourGlass();

        real34Compare(&counter, &loopTo, &compare);
        real34Compare(&loopStep, const34_0, &sign);
        real34Multiply(&compare, &sign, &compare);
        finished = real34ToInt32(&compare);                       //0 means equal
        if(finished > 0) {
          break;
        }

        if(getRegisterDataType(REGISTER_X) != dtReal34) {
          reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
        }
        real34Copy(&counter, REGISTER_REAL34_DATA(REGISTER_X));
        fnFillStack(NOPARAM);

        dynamicMenuItem = -1;
        execProgram(label);
        if(lastErrorCode != ERROR_NONE) {
          break;
        }

        if(getFlag(FLAG_CPXRES) && (getRegisterDataType(REGISTER_X) == dtComplex34 || !realIsZero(&resultRi))) {
            changedOverToComplex = true;     //Only latch over to complex operation if CPXRES is true, as well as either sum or new f(n) is complex
        }

        if(!changedOverToComplex) {
          fnToReal(NOPARAM);
          if(lastErrorCode != ERROR_NONE) {
            break;
          }
          real34ToReal(REGISTER_REAL34_DATA(REGISTER_X), &resultX); //Result accumulated
          if(prod) {
            realMultiply(&resultR, &resultX, &resultR, &ctxtReal75);
          }
          else {
            realAdd(&resultR, &resultX, &resultR, &ctxtReal75);
          }
        }
        else { //dtComplex34
          real34ToReal(REGISTER_REAL34_DATA(REGISTER_X), &resultX);  //Result accumulated
          real34ToReal(REGISTER_IMAG34_DATA(REGISTER_X), &resultXi); //Result accumulated
          if(prod) {
            mulComplexComplex(&resultR, &resultRi, &resultX, &resultXi, &resultR, &resultRi, &ctxtReal75);
          }
          else {
            realAdd(&resultR, &resultX, &resultR, &ctxtReal75);
            realAdd(&resultRi, &resultXi, &resultRi, &ctxtReal75);
          }
        }



        #if defined(VERBOSE_COUNTER)
          printf(">>> Fin: %d, Cpx: %d ", finished, changedOverToComplex);
          printReal34ToConsole(&counter," Cnt: ", " ");
          printRealToConsole(&resultX," X: ", " ");
          if(changedOverToComplex) {
            printRealToConsole(&resultXi," Xi: ", " ");
          }
          printRealToConsole(&resultR," SUM: ", "");
          if(changedOverToComplex) {
            printRealToConsole(&resultRi," SUMii: ", " ");
          }
          printf("\n");
        #endif // VERBOSE_COUNTER

        real34Add(&counter, &loopStep, &counter);
        if(printHalfSecUpdate_Integer(timed, "Loop: ",loop--, halfSec_clearZ, halfSec_clearT, halfSec_disp)) { ; //timed
          showProgressReal(&resultR, &resultRi, changedOverToComplex);
        }

        if(keyWaiting()) {
          showString("key Waiting ...", &standardFont, 16, Y_POSITION_OF_REGISTER_T_LINE, vmNormal, false, false);
          printHalfSecUpdate_Integer(force+1, "Interrupted: ", loop, halfSec_clearZ, halfSec_clearT, halfSec_disp);
          abort = true;
        }

        if(finished == 0 || abort) {
          break;
        }
      } //WHILE


      if(lastErrorCode == ERROR_NONE) {
        if(!changedOverToComplex) {
          if(getRegisterDataType(REGISTER_X) != dtReal34) {
            reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
          }
          convertRealToReal34ResultRegister(&resultR, REGISTER_X);
        }
        else {
          if(getRegisterDataType(REGISTER_X) != dtComplex34) {
            reallocateRegister(REGISTER_X, dtComplex34, COMPLEX34_SIZE_IN_BLOCKS, amNone);
          }
          convertRealToReal34ResultRegister(&resultR, REGISTER_X);
          convertRealToImag34ResultRegister(&resultRi, REGISTER_X);
        }
      adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
      }
      else {
        displayCalcErrorMessage(lastErrorCode, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          sprintf(errorMessage, "Error while calculating");
          moreInfoOnError("In function _programmableSumProd:", errorMessage, NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
        fnUndo(0);
      }

      temporaryInformation = TI_NO_INFO;
      if(programRunStop == PGM_WAITING) {
        programRunStop = PGM_STOPPED;
      }
    } //MAIN IF
    if((--currentSolverNestingDepth) == 0) {
      clearSystemFlag(FLAG_SOLVING);
    }
    hourGlassIconEnabled = false;
    showHideHourGlass();

    if(abort) {
    printHalfSecUpdate_Integer(force+0, "Loop aborted: ",loop, halfSec_clearZ, halfSec_clearT, halfSec_disp);
    } else {
    printHalfSecUpdate_Integer(force+0, "Loop complete: ",loop, halfSec_clearZ, halfSec_clearT, halfSec_disp);
    }
  }



  static void _checkArgument(uint16_t label, bool_t prod) {
    if(label >= FIRST_LABEL && label <= LAST_LABEL) {
      _programmableSumProd(label, prod);
    }
    else if(label >= REGISTER_X && label <= REGISTER_T) {
      // Interactive mode
      char buf[4];
      switch(label) {
        case REGISTER_X: {
          buf[0] = 'X';
          break;
        }
        case REGISTER_Y: {
          buf[0] = 'Y';
          break;
        }
        case REGISTER_Z: {
          buf[0] = 'Z';
          break;
        }
        case REGISTER_T: {
          buf[0] = 'T';
          break;
        }
        default: { /* unlikely */
          buf[0] = 0;
        }
      }
      buf[1] = 0;
      label = findNamedLabel(buf);
      if(label == INVALID_VARIABLE) {
        displayCalcErrorMessage(ERROR_LABEL_NOT_FOUND, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          sprintf(errorMessage, "string '%s' is not a named label", buf);
          moreInfoOnError("In function _checkArgument:", errorMessage, NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      }
      else {
        _programmableSumProd(label, prod);
      }
    }
    else {
      displayCalcErrorMessage(ERROR_OUT_OF_RANGE, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        sprintf(errorMessage, "unexpected parameter %u", label);
        moreInfoOnError("In function _checkArgument:", errorMessage, NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    }
  }
#endif // !TESTSUITE_BUILD

void fnProgrammableSum(uint16_t label) {
  #if !defined(TESTSUITE_BUILD)
    _checkArgument(label, false);
  #endif // !TESTSUITE_BUILD
}

void fnProgrammableProduct(uint16_t label) {
  #if !defined(TESTSUITE_BUILD)
    _checkArgument(label, true);
  #endif // !TESTSUITE_BUILD
}

File: zig_src/solver/sumprod.zig
// SPDX-License-Identifier: GPL-3.0-only
const consts = abi.constants;
//
// Zig owner for src/c47/solver/sumprod.c: the programmable Sigma / Pi
// (fnProgrammableSum / fnProgrammableProduct). This file is UNCOVERED by the
// testSuite (no test gate) - it is verified by build only. Faithful line-by-line
// translation preserving the exact order of every real_t / real34_t operation.
//
// The bridge (sumprod_legacy.c, removed) renamed fnProgrammableSum ->
// z47_solver_fnProgrammableSum and fnProgrammableProduct ->
// z47_solver_fnProgrammableProduct; those are the symbols solve.zig's
// dispatcher calls. We export them with those names. The static
// _programmableSumProd / _checkArgument become private. showProgressReal is also
// defined in sumprod.c and is used by graph.c, so it is re-exported here.
//
// The host-only progress display (showProgressReal body, guarded by
// ENABLE_SOLVER_PROGRESS == 1) has no effect on the computed stack result; the
// underlying control-flow functions (progressHalfSecUpdate_Integer, checkHalfSec)
// are still called so programRunStop / flag side effects are preserved, but the
// actual screen drawing is reduced to a no-op (same precedent as the prime
// owner's _showProgress). The VERBOSE_COUNTER debug blocks are #undef'd in the C
// and are omitted. EXTRA_INFO_ON_CALC_ERROR sprintf hints become fixed
// moreInfoOnError strings (no-op under TESTSUITE/DMCP).

const runtime = @import("solve_runtime.zig");

// DECNUMDIGITS=75, DECDPUN=3 => DECNUMUNITS=ceil(75/3)=25; decNumberUnit=u16.
const abi = @import("abi"); // L1 shared bindings (REPORT-23 §5)
const real_t = abi.Real;
const real34_t = abi.Real34;
const realContext_t = abi.RealContext;

const calcRegister_t = runtime.calcRegister_t;
const bool_t = bool;

// GMP long integer (mpz). The linkable symbols are __gmpz_*.
const mp_limb_t = usize;
const mpz_struct = abi.Mpz;
const longInteger_t = [1]mpz_struct;

extern fn __gmpz_init(op: *mpz_struct) void;
extern fn __gmpz_clear(op: *mpz_struct) void;
extern fn __gmpz_fdiv_ui(op: *const mpz_struct, d: c_ulong) c_ulong;

inline fn longIntegerInit(op: *mpz_struct) void {
    __gmpz_init(op);
}
inline fn longIntegerFree(op: *mpz_struct) void {
    __gmpz_clear(op);
}
inline fn longIntegerModuloUInt(op: *const mpz_struct, u: u32) u32 {
    return @truncate(__gmpz_fdiv_ui(op, u));
}

// ---------------------------------------------------------------------------
// defines.h values (verified)
// ---------------------------------------------------------------------------
const REGISTER_X: calcRegister_t = 100;
const REGISTER_Y: calcRegister_t = 101;
const REGISTER_Z: calcRegister_t = 102;
const REGISTER_T: calcRegister_t = 103;
const ERR_REGISTER_LINE: calcRegister_t = REGISTER_Z;

const FIRST_LABEL: u16 = 2200; // INVALID_VARIABLE=2199 precedes FIRST_LABEL; the //2044 C comment is stale
const LAST_LABEL: u16 = 6999;
const INVALID_VARIABLE: u16 = 2199;

const ERROR_NONE: u8 = 0;
const ERROR_LABEL_NOT_FOUND: u8 = 6;
const ERROR_OUT_OF_RANGE: u8 = 8;
const ERROR_BAD_INPUT: u8 = 53;

const FLAG_SOLVING: u32 = 0xc026;
const FLAG_CPXRES: u16 = 0x8004;

const dtReal34: u32 = 1;
const dtComplex34: u32 = 2;
const amNone: u32 = 5;
const NOPARAM: u16 = 9876;

const DEC_ROUND_DOWN: c_int = 5; // round towards 0 (truncate)

const TI_NO_INFO: u8 = 0;
const PGM_WAITING: u8 = 2;
const PGM_STOPPED: u8 = 0;

// screen.h: timed=0, force=1, halfSec_clear*/disp=true.
const timed: u8 = 0;
const force: u8 = 1;
const halfSec_clearZ: bool = true;
const halfSec_clearT: bool = true;
const halfSec_disp: bool = true;

// ---------------------------------------------------------------------------
// Globals
// ---------------------------------------------------------------------------
extern var lastErrorCode: u8;
extern var temporaryInformation: u8;
extern var programRunStop: u8;
extern var currentKeyCode: u8;
extern var dynamicMenuItem: i16;
extern var currentSolverNestingDepth: u16;

extern var ctxtReal75: realContext_t;
extern var ctxtReal34: realContext_t;

// ---------------------------------------------------------------------------
// Constants blob accessors
// ---------------------------------------------------------------------------

inline fn const_0() *align(1) const real_t {
    return consts.c1708();
}
inline fn const_1() *align(1) const real_t {
    return consts.c4856();
}
inline fn const34_0() *align(1) const real34_t {
    return consts.q16200();
}

// ---------------------------------------------------------------------------
// decNumber primitives / real_t macros
// ---------------------------------------------------------------------------
extern fn decNumberCopy(res: *real_t, source: *align(1) const real_t) *real_t;
extern fn decNumberAdd(res: *real_t, op1: *align(1) const real_t, op2: *align(1) const real_t, ctxt: *realContext_t) *real_t;
extern fn decNumberMultiply(res: *real_t, op1: *align(1) const real_t, op2: *align(1) const real_t, ctxt: *realContext_t) *real_t;

// decNumberIsZero is a decNumber.h macro (DECSPECIAL = 0x70):
//   (*lsu == 0 && digits == 1 && (bits & 0x70) == 0)
inline fn decNumberIsZero(dn: *align(1) const real_t) bool {
    return dn.lsu[0] == 0 and dn.digits == 1 and (dn.bits & 0x70) == 0;
}

inline fn realCopy(source: *align(1) const real_t, destination: *real_t) void {
    _ = decNumberCopy(destination, source);
}
inline fn realAdd(op1: *align(1) const real_t, op2: *align(1) const real_t, res: *real_t, ctxt: *realContext_t) void {
    _ = decNumberAdd(res, op1, op2, ctxt);
}
inline fn realMultiply(op1: *align(1) const real_t, op2: *align(1) const real_t, res: *real_t, ctxt: *realContext_t) void {
    _ = decNumberMultiply(res, op1, op2, ctxt);
}
inline fn realIsZero(source: *const real_t) bool {
    return decNumberIsZero(source);
}
extern fn realSetZero(value: *real_t) void;
extern fn mulComplexComplex(f1r: *const real_t, f1i: *const real_t, f2r: *const real_t, f2i: *const real_t, pr: *real_t, pi: *real_t, real_context: *realContext_t) void;

// ---------------------------------------------------------------------------
// real34_t macros
// ---------------------------------------------------------------------------
extern fn decimal128ToNumber(source: *align(1) const real34_t, destination: *real_t) *real_t;
extern fn decQuadAdd(res: *align(1) real34_t, op1: *align(1) const real34_t, op2: *align(1) const real34_t, ctxt: *realContext_t) *align(1) real34_t;
extern fn decQuadSubtract(res: *align(1) real34_t, op1: *align(1) const real34_t, op2: *align(1) const real34_t, ctxt: *realContext_t) *align(1) real34_t;
extern fn decQuadMultiply(res: *align(1) real34_t, op1: *align(1) const real34_t, op2: *align(1) const real34_t, ctxt: *realContext_t) *align(1) real34_t;
extern fn decQuadDivide(res: *align(1) real34_t, op1: *align(1) const real34_t, op2: *align(1) const real34_t, ctxt: *realContext_t) *align(1) real34_t;
extern fn decQuadCompare(res: *align(1) real34_t, op1: *align(1) const real34_t, op2: *align(1) const real34_t, ctxt: *realContext_t) *align(1) real34_t;
extern fn decQuadIsZero(source: *align(1) const real34_t) bool;
extern fn decQuadToInt32(source: *align(1) const real34_t, ctxt: *realContext_t, mode: c_int) i32;
extern fn convertReal34ToLongInteger(real34: *align(1) const real34_t, lgInt: *mpz_struct, mode: c_int) void;

inline fn real34ToReal(source: *align(1) const real34_t, destination: *real_t) void {
    _ = decimal128ToNumber(source, destination);
}
inline fn real34Copy(source: *align(1) const real34_t, destination: *align(1) real34_t) void {
    const src: *align(1) const [2]u64 = @ptrCast(source);
    const dst: *align(1) [2]u64 = @ptrCast(destination);
    dst[0] = src[0];
    dst[1] = src[1];
}
inline fn real34Add(op1: *align(1) const real34_t, op2: *align(1) const real34_t, res: *align(1) real34_t) void {
    _ = decQuadAdd(res, op1, op2, &ctxtReal34);
}
inline fn real34Subtract(op1: *align(1) const real34_t, op2: *align(1) const real34_t, res: *align(1) real34_t) void {
    _ = decQuadSubtract(res, op1, op2, &ctxtReal34);
}
inline fn real34Multiply(op1: *align(1) const real34_t, op2: *align(1) const real34_t, res: *align(1) real34_t) void {
    _ = decQuadMultiply(res, op1, op2, &ctxtReal34);
}
inline fn real34Divide(op1: *align(1) const real34_t, op2: *align(1) const real34_t, res: *align(1) real34_t) void {
    _ = decQuadDivide(res, op1, op2, &ctxtReal34);
}
inline fn real34Compare(op1: *align(1) const real34_t, op2: *align(1) const real34_t, res: *align(1) real34_t) void {
    _ = decQuadCompare(res, op1, op2, &ctxtReal34);
}
inline fn real34IsZero(source: *align(1) const real34_t) bool {
    return decQuadIsZero(source);
}
inline fn real34ToInt32(source: *align(1) const real34_t) i32 {
    return decQuadToInt32(source, &ctxtReal34, DEC_ROUND_DOWN);
}

extern fn real34CompareEqual(number1: *align(1) const real34_t, number2: *align(1) const real34_t) bool;
extern fn real34CompareGreaterThan(number1: *align(1) const real34_t, number2: *align(1) const real34_t) bool;
extern fn real34CompareLessThan(number1: *align(1) const real34_t, number2: *align(1) const real34_t) bool;
extern fn real34CompareGreaterEqual(number1: *align(1) const real34_t, number2: *align(1) const real34_t) bool;
extern fn real34CompareLessEqual(number1: *align(1) const real34_t, number2: *align(1) const real34_t) bool;

// ---------------------------------------------------------------------------
// Register / stack / program externs
// ---------------------------------------------------------------------------
const REAL34_SIZE_IN_BYTES: usize = 16;
extern fn getRegisterDataPointer(reg: calcRegister_t) ?*anyopaque;
const registerReal34Ptr = abi.registerReal34;
const registerImag34Ptr = abi.registerImag34;

extern fn getRegisterDataType(reg: calcRegister_t) u32;
extern fn reallocateRegister(regist: calcRegister_t, data_type: u32, data_len: u32, tag: u32) void;
extern fn convertRealToReal34ResultRegister(real: *const real_t, dest: calcRegister_t) void;
extern fn convertRealToImag34ResultRegister(real: *const real_t, dest: calcRegister_t) void;

extern fn getFlag(flag: u16) bool;
extern fn setSystemFlag(sf: c_uint) void;
extern fn clearSystemFlag(sf: c_uint) void;
extern fn fnToReal(unused: u16) void;
extern fn fnDrop(unused: u16) void;
extern fn fnFillStack(unused: u16) void;
extern fn execProgram(label: u16) void;
extern fn adjustResult(res: calcRegister_t, drop_y: bool, set_cpx_res: bool, op1: calcRegister_t, op2: calcRegister_t, op3: calcRegister_t) void;
extern fn displayCalcErrorMessage(error_code: u8, err_message_register_line: calcRegister_t, err_register_line: calcRegister_t) void;
extern fn moreInfoOnError(m1: [*:0]const u8, m2: ?[*:0]const u8, m3: ?[*:0]const u8, m4: ?[*:0]const u8) void;
extern fn findNamedLabel(label_name: [*:0]const u8) calcRegister_t;
extern fn letteredRegisterName(regist: calcRegister_t) u8;

extern fn checkHalfSec() bool;
extern fn progressHalfSecUpdate_Integer(mode: u8, txt: [*:0]const u8, loop: i32, clearZ: bool, clearT: bool, disp: bool) bool;

// ===========================================================================
// showProgressReal: host-only progress display. The drawing is reduced to a
// no-op (no effect on the computed result); still exported because graph.c
// calls it. (ENABLE_SOLVER_PROGRESS == 1 body omitted.)
// ===========================================================================
pub export fn showProgressReal(a: *const real_t, ai: *real_t, cpx: bool_t) linksection(runtime.code_section) callconv(.c) void {
    _ = a;
    _ = ai;
    _ = cpx;
}

// ===========================================================================
// _programmableSumProd
// ===========================================================================
fn _programmableSumProd(label: u16, prod: bool_t) linksection(runtime.code_section) void {
    currentKeyCode = 255;
    var loop: i32 = 0;
    var finished: i16 = 0;
    var resultX: real_t = undefined;
    var resultXi: real_t = undefined;
    var resultR: real_t = undefined;
    var resultRi: real_t = undefined;
    var loopStep: real34_t = undefined;
    var loopTo: real34_t = undefined;
    var counter: real34_t = undefined;
    var compare: real34_t = undefined;
    var sign: real34_t = undefined;
    var rLoop: real34_t = undefined;
    var changedOverToComplex: bool_t = false;
    var iLoop: longInteger_t = undefined;

    fnToReal(NOPARAM);
    real34Copy(registerReal34Ptr(REGISTER_X), &loopStep);
    fnDrop(NOPARAM);
    fnToReal(NOPARAM);
    real34Copy(registerReal34Ptr(REGISTER_X), &loopTo);
    fnDrop(NOPARAM);
    fnToReal(NOPARAM);
    real34Copy(registerReal34Ptr(REGISTER_X), &counter); // Loopfrom
    realCopy(if (prod) const_1() else const_0(), &resultR); // Initialize real accumulator
    realSetZero(&resultRi); // Initialize complex accumulator

    real34Subtract(&loopTo, &counter, &rLoop); // calculate the remaining iteration counter
    if (!real34IsZero(&loopStep)) {
        real34Divide(&rLoop, &loopStep, &rLoop);
    }
    // iLoop is initialised by convertReal34ToLongInteger; do not double-init (leak).
    convertReal34ToLongInteger(&rLoop, &iLoop[0], DEC_ROUND_DOWN);
    loop = @bitCast(longIntegerModuloUInt(&iLoop[0], @bitCast(@as(i32, 0x7FFFFFFF))));
    longIntegerFree(&iLoop[0]);

    if (!real34CompareEqual(&loopTo, &counter) and
        (real34IsZero(&loopStep) or
            (real34CompareGreaterThan(&loopTo, &counter) and real34CompareLessEqual(&loopStep, const34_0())) or
            (real34CompareLessThan(&loopTo, &counter) and real34CompareGreaterEqual(&loopStep, const34_0()))))
    {
        displayCalcErrorMessage(ERROR_BAD_INPUT, ERR_REGISTER_LINE, REGISTER_X);
        moreInfoOnError("In function _programmableSumProd:", "Counter will not count to destination", null, null);
    } else {
        currentSolverNestingDepth += 1;
        setSystemFlag(FLAG_SOLVING);

        while (lastErrorCode == ERROR_NONE) {
            loop -= 1;
            if (checkHalfSec()) {
                if (progressHalfSecUpdate_Integer(timed, "Loop: ", loop, halfSec_clearZ, halfSec_clearT, halfSec_disp)) {
                    showProgressReal(&resultR, &resultRi, changedOverToComplex);
                }
            }

            real34Compare(&counter, &loopTo, &compare);
            real34Compare(&loopStep, const34_0(), &sign);
            real34Multiply(&compare, &sign, &compare);
            finished = @truncate(real34ToInt32(&compare)); // 0 means equal
            if (finished > 0) {
                break;
            }

            if (getRegisterDataType(REGISTER_X) != dtReal34) {
                reallocateRegister(REGISTER_X, dtReal34, 0, amNone);
            }
            real34Copy(&counter, registerReal34Ptr(REGISTER_X));
            fnFillStack(NOPARAM);

            dynamicMenuItem = -1;
            execProgram(label);
            if (lastErrorCode != ERROR_NONE) {
                break;
            }

            if (getFlag(FLAG_CPXRES) and (getRegisterDataType(REGISTER_X) == dtComplex34 or !realIsZero(&resultRi))) {
                changedOverToComplex = true; // Only latch over to complex operation if CPXRES is true, as well as either sum or new f(n) is complex
            }

            if (!changedOverToComplex) {
                fnToReal(NOPARAM);
                if (lastErrorCode != ERROR_NONE) {
                    break;
                }
                real34ToReal(registerReal34Ptr(REGISTER_X), &resultX); // Result accumulated
                if (prod) {
                    realMultiply(&resultR, &resultX, &resultR, &ctxtReal75);
                } else {
                    realAdd(&resultR, &resultX, &resultR, &ctxtReal75);
                }
            } else { // dtComplex34
                real34ToReal(registerReal34Ptr(REGISTER_X), &resultX); // Result accumulated
                real34ToReal(registerImag34Ptr(REGISTER_X), &resultXi); // Result accumulated
                if (prod) {
                    mulComplexComplex(&resultR, &resultRi, &resultX, &resultXi, &resultR, &resultRi, &ctxtReal75);
                } else {
                    realAdd(&resultR, &resultX, &resultR, &ctxtReal75);
                    realAdd(&resultRi, &resultXi, &resultRi, &ctxtReal75);
                }
            }

            real34Add(&counter, &loopStep, &counter);

            if (finished == 0) {
                break;
            }
        } // WHILE

        if (lastErrorCode == ERROR_NONE) {
            if (!changedOverToComplex) {
                if (getRegisterDataType(REGISTER_X) != dtReal34) {
                    reallocateRegister(REGISTER_X, dtReal34, 0, amNone);
                }
                convertRealToReal34ResultRegister(&resultR, REGISTER_X);
            } else {
                if (getRegisterDataType(REGISTER_X) != dtComplex34) {
                    reallocateRegister(REGISTER_X, dtComplex34, 0, amNone);
                }
                convertRealToReal34ResultRegister(&resultR, REGISTER_X);
                convertRealToImag34ResultRegister(&resultRi, REGISTER_X);
            }
            adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
        } else {
            displayCalcErrorMessage(lastErrorCode, ERR_REGISTER_LINE, REGISTER_X);
            moreInfoOnError("In function _programmableSumProd:", "Error or exit while calculating", null, null);
        }

        temporaryInformation = TI_NO_INFO;
        if (programRunStop == PGM_WAITING) {
            programRunStop = PGM_STOPPED;
        }
    } // MAIN IF

    currentSolverNestingDepth -= 1;
    if (currentSolverNestingDepth == 0) {
        clearSystemFlag(FLAG_SOLVING);
    }
}

// ===========================================================================
// _checkArgument
// ===========================================================================
fn _checkArgument(label_in: u16, prod: bool_t) linksection(runtime.code_section) void {
    var label = label_in;
    if (FIRST_LABEL <= label and label <= LAST_LABEL) {
        _programmableSumProd(label, prod);
    } else if (REGISTER_X <= @as(calcRegister_t, @intCast(label)) and @as(calcRegister_t, @intCast(label)) <= REGISTER_T) {
        // Interactive mode
        var buf: [2]u8 = undefined;
        buf[0] = letteredRegisterName(@intCast(label));
        buf[1] = 0;
        label = @bitCast(@as(i16, @truncate(findNamedLabel(@ptrCast(&buf[0])))));
        if (label == INVALID_VARIABLE) {
            displayCalcErrorMessage(ERROR_LABEL_NOT_FOUND, ERR_REGISTER_LINE, REGISTER_X);
            moreInfoOnError("In function _checkArgument:", "string is not a named label", null, null);
        } else {
            _programmableSumProd(label, prod);
        }
    } else {
        displayCalcErrorMessage(ERROR_OUT_OF_RANGE, ERR_REGISTER_LINE, REGISTER_X);
        moreInfoOnError("In function _checkArgument:", "unexpected parameter", null, null);
    }
}

pub export fn z47_solver_fnProgrammableSum(label: u16) linksection(runtime.code_section) callconv(.c) void {
    _checkArgument(label, false);
}

pub export fn z47_solver_fnProgrammableProduct(label: u16) linksection(runtime.code_section) callconv(.c) void {
    _checkArgument(label, true);
}
